components/TextFields/src/MDCTextField.m

```diff
@@ -0,0 +1,530 @@
+  if (self) {
+    self.leftView = leftView;
+  }
+  return self;
+}
+
+- (void)dealloc {
+  NSNotificationCenter *defaultCenter = [NSNotificationCenter defaultCenter];
+  [defaultCenter removeObserver:self name:UITextFieldTextDidBeginEditingNotification object:self];
+  [defaultCenter removeObserver:self name:UITextFieldTextDidChangeNotification object:self];
+}
+
+- (void)encodeWithCoder:(NSCoder *)aCoder {
+  [super encodeWithCoder:aCoder];
+  [aCoder encodeObject:self.clearButtonColor forKey:MDCTextFieldClearButtonColorKey];
+  [aCoder encodeObject:self.clearButtonImage forKey:MDCTextFieldClearButtonImageKey];
+  [aCoder encodeConditionalObject:self.coordinator forKey:MDCTextFieldCoordinatorKey];
+}
+
+- (instancetype)copyWithZone:(NSZone *)zone {
+  MDCTextField *copy = [[[self class] alloc] init];
+
+  copy.clearButtonColor = self.clearButtonColor.copy;
+  copy.clearButtonImage = self.clearButtonImage.copy;
+
+  // Just a pointer value copies.
+  copy.coordinator = self.coordinator;
+
+  return copy;
+}
+
+- (void)commonMDCTextFieldInitialization {
+
+  [super setBorderStyle:UITextBorderStyleNone];
+
+  // Set the clear button color to black with 54% opacity.
+  [self setClearButtonColor:[UIColor colorWithWhite:0 alpha:[MDCTypography captionFontOpacity]]];
+
+  NSNotificationCenter *defaultCenter = [NSNotificationCenter defaultCenter];
+  [defaultCenter addObserver:self
+                    selector:@selector(textFieldDidBeginEditing:)
+                        name:UITextFieldTextDidBeginEditingNotification
+                      object:self];
+  [defaultCenter addObserver:self
+                    selector:@selector(textFieldDidChange:)
+                        name:UITextFieldTextDidChangeNotification
+                      object:self];
+}
+
+#pragma mark - Clear Button Image
+
+- (UIImage *)drawnClearButtonImage:(CGSize)size {
+  if (CGSizeEqualToSize(size, CGSizeZero)) {
+    size = CGSizeMake(MDCClearButtonImageSquareSize, MDCClearButtonImageSquareSize);
+  }
+  CGFloat scale = [UIScreen mainScreen].scale;
+  CGRect bounds = CGRectMake(0, 0, size.width * scale, size.height * scale);
+  UIGraphicsBeginImageContextWithOptions(bounds.size, false, scale);
+  [self.clearButtonColor setFill];
+  [MDCPathForClearButtonImageFrame(bounds) fill];
+  UIImage *image = UIGraphicsGetImageFromCurrentImageContext();
+  UIGraphicsEndImageContext();
+
+  return image;
+}
+
+#pragma mark - Properties Implementation
+
+- (void)setClearButtonColor:(UIColor *)clearButtonColor {
+  if (_clearButtonColor != clearButtonColor) {
+    _clearButtonColor = clearButtonColor;
+    self.clearButtonImage = [self drawnClearButtonImage:self.clearButtonImage.size];
+  }
+}
+
+- (BOOL)hidesPlaceholderOnInput {
+  return _coordinator.hidesPlaceholderOnInput;
+}
+
+- (void)setHidesPlaceholderOnInput:(BOOL)hidesPlaceholderOnInput {
+  _coordinator.hidesPlaceholderOnInput = hidesPlaceholderOnInput;
+}
+
+- (UIButton *)internalClearButton {
```

Comment:
This method is shady.
Since we are required to do this, can you add a comment at the top:
// Returns the internal clear button.  
//We have to do X since it isn't exposed by UIKit.
Is there any way we can leave UITextField's clearButton to Never and implement our own?